File: cubes.py
#!/bin/ipython3

# Constants ---------------------------------------------------

# Direction deffinitions
CW = 0
CCW = 1

# For easier reading when referenceing coordinates
X = 0
Y = 1

# Map of directions to coorsponding numbers
N 	= 0
NE 	= 1
E 	= 2
SE 	= 3
S 	= 4
SW 	= 5
W 	= 6
NW 	= 7

# Lists of directions for iteration
DA = [N,E,S,W]
DB = [N,NE,E,SE,S,SW,W,NW]
DC = [(0,-1), (1,-1), (1,0), (1,1), (0,1), (-1,1), (-1,0), (-1,-1)]

pivotTable = [[W,N],[N,E],[E,S],[S,W]]

# pivotCheckTable = [[[(x1,y1+1),(x2,y2+1),(x2,y2)],[(x1+1,y1),(x2+1,y2),(x2,y2)]],
# 				   	   [[(x1-1,y1),(x2-1,y2),(x2,y2)],[(x1,y1+1),(x2,y2+1),(x2,y2)]],
# 				   	   [[(x1,y1-1),(x2,y2-1),(x2,y2)],[(x1-1,y1),(x2-1,y2),(x2,y2)]],
# 				   	   [[(x1+1,y1),(x2+1,y2),(x2,y2)],[(x1,y1-1),(x2,y2-1),(x2,y2)]]]

pivotCheckTable = [[[(0,1),(0,1),(0,0)],[(1,0),(1,0),(0,0)]],
			   	   [[(-1,0),(-1,0),(0,0)],[(0,1),(0,1),(0,0)]],
			   	   [[(0,-1),(0,-1),(0,0)],[(-1,0),(-1,0),(0,0)]],
			   	   [[(1,0),(1,0),(0,0)],[(0,-1),(0,-1),(0,0)]]]

squaresList = []


# Functions, Procedures, Classes & Methods --------------------------------------------------------

class coordinate(tuple):
	def __add__(self, direction):
		return coordinate((self[X] + DC[direction][X], self[Y] + DC[direction][Y]))

# Returns opposite direction of orientation
def opposite(orientation):
	return((orientation+4)%8)

# Increments orientation in direction by 45 degrees num times
def clock(orientation, direction, num = 1):
	d = (1 if direction == CW else -1)
	for n in range(num):
		orientation = (orientation + d) % 8
	return orientation

def squareAt(coord):
	r = False
	for square in squaresList:
		if square.location == coord:
			r = True
			continue
	return r


def moveResult(location, direction):
	return (location[X]+DC[direction][X],location[Y]+DC[direction][Y])

def pivotResult(location, corner, direction):
	return moveResult(location, pivotTable[corner][direction])
	

def canPivot(location, corner, direction):
	x1 = location[X]
	y1 = location[Y]
	pivotLocation = pivotResult(location,corner,direction)
	x2 = pivotLocation[X]
	y2 = pivotLocation[Y]
	
	# Get the locations where squares whould interfere with this pivot	# 
	# Get list off ofsets from pre computed table
	cl = pivotCheckTable[corner][direction]
	# Add offsets to coordinates to find list of coordinates to check
	checkList = [(x1+cl[0][0],y1+cl[0][1]),(x2+cl[1][0],y2+cl[1][1]),(x2+cl[2][0],y2+cl[2][1])]
	
	# Check interfering locations
	for coord in checkList:
		if squareAt(coord):
			return False
	
	return True



class Square(object):
	def __init__(self, x = -1, y = -1):
			# self.fill = "grey"
			#self.master = master
			#self.connections = {}
			self.drawings = []
			self.location = (x,y)
			self.orientation = N
			squaresList.append(self)
			
	# Pivot self in given direction about specified corner
	def pivot(self, corner, direction):
		if canPivot(self.location,corner,direction):
			pivotLocation = pivotResult(self.location, corner, direction)
			adjecentList = [squareAt((pivotLocation[X]+DC[d][X],pivotLocation[Y]+DC[d][Y])) for d in DA if (pivotLocation[X]+DC[d][X],pivotLocation[Y]+DC[d][Y]) != self.location]
			diagonalList = [squareAt((pivotLocation[X]+DC[d+1][X],pivotLocation[Y]+DC[d+1][Y])) for d in DA if (pivotLocation[X]+DC[d+1][X],pivotLocation[Y]+DC[d+1][Y]) != self.location]
			if True not in adjecentList:
				if True in diagonalList:
					# Check if we can pivot 90 degrees further to reach another square
					pCorner = (corner+1)%4 if direction == CW else (corner-1)%4
					if canPivot(pivotLocation, pCorner, direction):
						self.move(pivotTable[corner][direction])
						self.rotate(direction)
						self.move(pivotTable[pCorner][direction])
						self.rotate(direction)
						return
			else:
				self.move(pivotTable[corner][direction])
				self.rotate(direction)


		

	def rotate(self, direction):
		self.orientation = clock(self.orientation, direction, 2)

	
	
	def move(self, direction):
		self.location = (self.location[X]+DC[direction][X],self.location[Y]+DC[direction][Y])
			

	

File: cubes.js
// Constants ---------------------------------------------------

// Direction deffinitions
var CW = 1;
var CCW = -1;

// For easier reading when referenceing coordinates
var X = 0;
var Y = 1;

// Map of directions to coorsponding numbers
var N 	= 0;
var NE 	= 1;
var E 	= 2;
var SE 	= 3;
var S 	= 4;
var SW 	= 5;
var W 	= 6;
var NW 	= 7;

// Lists of directions for iteration
var DA = [N,E,S,W];
var DB = [N,NE,E,SE,S,SW,W,NW];
var DC = [[0,1], [1,1], [1,0], [1,-1], [0,-1], [-1,-1], [-1,0], [-1,1]];

// Size of window
var SIZE = 512;
// Size of squares
var SCALE = 16;

// Derived info about window and grid
var GSIZE = (SIZE/SCALE)-1;
var GCENTER = (GSIZE/2);
var X_SCALE = SCALE;
var Y_SCALE = SCALE;

var squaresList = new Array();
// squaresCoords = [[0 for i in range(GSIZE)] for j in range(GSIZE)]
var squaresCoords = new Array();

var MASTER = 0;


// Returns opposite direction of orientation
function opposite(orientation) {
	return((orientation+4)%8);
}

// Increments orientation in direction num times
function clock(orientation, direction, num){
	while (num > 0) {
		if (direction == CW) {
			if (orientation == NW) {
				orientation = N;
			} else {
				orientation = orientation+1;
			}
		} else if( direction == CCW) {
			if (orientation == N) {
				orientation = NW;
			} else {
				orientation = orientation-1;
			}
		}
		num -= 1;
	}
	return orientation;
}

function coordAddDir(a, b){
	return [a[X] + DC[b][X], a[Y] + DC[b][Y]];
}



function Square(parent, parentDir, master) { 
	this.fill = "grey";
	this.master = master;
	this.connections = new Array();
	this.orientation = N;
	this.adjNum = 0;

	if (this.master) {
		MASTER = this;
		this.coord = [0, 0];
		squaresList.push(this);
		squaresCoords[this.coord] = this;
		getConnections(this);
		// this.getConnected();
	} else {
		this.coord = coordAddDir(parent.coord, parentDir);
		this.connections[opposite(parentDir)] = parent;
		parent.connections[parentDir] = this;

		getConnections(this);
		// this.getConnected();

		squaresCoords[this.coord] = this;
		squaresList.push(this);
		for (s in this.connections.keys()) {
			alert("test");
			this.connections[s].getConnections();
		}
	}
}

// Pivot self in given direction
Square.prototype.pivot = function(direction) {
	getConnections(this);
	for (var d = 0; d < DB.length; d+= 2) {
		if (this.connections[d]) {
			pivot = d;
		}
	}

	if (this.connections[opposite(pivot)] == null) {
		t = clock(pivot, -direction, 2);
		if (this.connections[t]) {
			pivot = t;
			t = clock(pivot, -direction, 2);
		}
	

		temp = coordAddDir(this.coord, t);

		if (squaresCoords[temp] == null) {
			if (squaresCoords[coordAddDir(temp, pivot)]) {
				if (squaresCoords[coordAddDir(temp, opposite(pivot))] == null) {
					if (move(this, coordAddDir(this.coord, t))){
						this.orientation = clock(this.orientation, direction, 2);
						return 1;
					}
				}
			} else {
				if (squaresCoords[coordAddDir(temp, t)]) {
					if (move(this, coordAddDir(this.coord, t))) {
						this.orientation = clock(this.orientation, direction, 2);
						return 1;
					}
				} else {
					t = clock(pivot, -direction, 1)
					if (squaresCoords[coordAddDir(temp, t)] == null) {
						if (move(this, coordAddDir(this.coord, t))) {
							this.orientation = clock(this.orientation, direction, 4);
							return 1;
						}
					}
				}
			}
		}
	}
	return 0
}

function move(square, newCoord) {
	var fail = 0;
	var oldCoord = square.coord;
	var conns = [];
	if (squaresCoords[newCoord] == null) {
		squaresCoords[square.coord] = null;
		square.coord = newCoord;
		squaresCoords[square.coord] = square;			

		for (var s = 0; s < squaresList.length; s++) {
			getConnections(squaresList[s])
			// squaresList[s].getConnected()
			if (squaresList[s].adjNum == 0) {
				fail = 1;
				break;
			}
		}

		if (fail) {
			squaresCoords[square.coord] == null
			square.coord = oldCoord;
			squaresCoords[square.coord] = square;
			for (var s = 0; s < squaresList.length; s++) {
				getConnections(squaresList[s])
				// squaresList[s].getConnected()
			}
			return 0;
		} else {
			return 1;
		}
	} else {
		return 0;
	}
}


// Gets the direcetions from self which have squares
function getConnections(square) {
	square.connections = new Array();
	square.adjNum = 0;
	for (var d = 0; d < DB.length; d++) {
		c = coordAddDir(square.coord, d);
		if (squaresCoords[c]) {
			square.connections[d] = squaresCoords[c];
			squaresCoords[c].connections[opposite(d)] = square;
			if ((d == 0) || (d == 2) || (d == 4) || (d == 6)) {
				square.adjNum += 1;
			}
		}
	}
}




function test() {
	var a = new Square(0, 0, 1)
	var b = new Square(MASTER, N)
	b.pivot(CW)


}